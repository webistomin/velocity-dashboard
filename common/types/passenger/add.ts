import { Request } from 'express';
import { IPassengerInterface } from './passenger-schema';

export interface IPassengerAddRequest extends Request {
  body: IPassengerInterface;
}
