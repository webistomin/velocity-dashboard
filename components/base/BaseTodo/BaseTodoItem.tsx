import { VueComponent } from 'types/vue-components';
import { Component, Prop, Emit } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import { add, compareAsc } from 'date-fns';

import { clientUrls } from 'common/urls/clientUrls';
import { IBaseTodo } from './BaseTodo';

export interface IBaseTodoItem extends IBaseTodo {
  onInput?: (event: Event) => void;
}

@Component({
  name: 'BaseTodoItem',
})
export default class BaseTodoItem extends VueComponent<IBaseTodoItem> {
  @Prop()
  private readonly id!: IBaseTodo['id'];

  @Prop()
  private readonly title!: IBaseTodo['title'];

  @Prop()
  private readonly dueDate!: IBaseTodo['dueDate'];

  @Prop()
  private readonly isDone!: IBaseTodo['isDone'];

  public get getDeadline(): string {
    const taskDate = new Date(this.dueDate);
    const today = new Date();
    const onTimeDate = add(new Date(), { days: 5 });

    /* Compare the two dates and return 1 if the first date is after the second,
     * -1 if the first date is before the second
     * or 0 if dates are equal. */
    if (compareAsc(today, taskDate) === 1) {
      return 'overdue';
    } else if (compareAsc(onTimeDate, taskDate) === 1) {
      return 'due-soon';
    }

    return 'on-time';
  }

  @Emit('input')
  public onInput(event: Event): Event {
    return event;
  }

  public render(): VNode {
    return (
      <li class={`list-item base-todo__item ${this.isDone ? 'base-todo__item_done' : ''}`}>
        <div class='base-todo__actions'>
          <input
            type='checkbox'
            class='base-todo__checkbox visually-hidden'
            id={this.id}
            onInput={this.onInput}
            value={this.id}
            checked={this.isDone}
            aria-label={`Toggle task with name: ${this.title}`}
          />
          <label for={this.id} class='base-todo__label label' />
        </div>
        <nuxt-link to={`${clientUrls.reminders}/${this.id}`} class='base-todo__content link'>
          <strong class='base-todo__name'>{this.title}</strong>
          <time class={`base-todo__date base-todo__date_deadline_${this.getDeadline}`}>
            Due {new Date(this.dueDate).toLocaleString()}
          </time>
        </nuxt-link>
      </li>
    );
  }
}
