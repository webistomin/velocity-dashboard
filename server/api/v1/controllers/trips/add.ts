import { Response } from 'express';
import HTTPStatuses from 'http-status-codes';

import Trip from 'server/models/trip/trip';
import { ITripAddRequest, ITripAddResponse } from 'common/types/trip/trip-add';

export default async (req: ITripAddRequest, res: Response<ITripAddResponse>) => {
  try {
    const data = req.body;
    const newTrip = new Trip();

    newTrip.passengerId = data.passengerId;
    newTrip.trip = data.trip;
    newTrip.paymentDetails = data.paymentDetails;
    newTrip.path = data.path;

    await newTrip.save();

    return res.status(HTTPStatuses.CREATED).json({
      success: true,
      message: 'Trip created',
    });
  } catch (error) {
    return res.status(HTTPStatuses.INTERNAL_SERVER_ERROR).json({
      success: false,
      message: error.message,
    });
  }
};
