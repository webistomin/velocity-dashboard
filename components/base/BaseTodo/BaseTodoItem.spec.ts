import { mount } from '@vue/test-utils';
import add from 'date-fns/add';
import sub from 'date-fns/sub';
import BaseTodoItem from './BaseTodoItem';

describe('BaseTodoItem', () => {
  const checkboxSelector = "[data-jest='base-todo__checkbox']";
  const itemSelector = "[data-jest='base-todo__item']";
  const dateSelector = "[data-jest='base-todo__date']";

  const factory = (propsData = {}, options = {}) => {
    return mount(BaseTodoItem, {
      propsData: {
        id: '1',
        title: 'Title',
        isDone: false,
        ...propsData,
      },
      stubs: ['nuxt-link'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Emit "input" event', () => {
    const wrapper = factory();
    const input = wrapper.find(checkboxSelector);
    input.trigger('input');
    expect(wrapper.emitted().input);

    const onInputMethod = wrapper.vm.onInput({} as Event);
    expect(onInputMethod).toEqual({});
  });

  it('Set done class is "isDone" prop is provided', () => {
    const expectedResult = ['base-todo__item_done'];
    const wrapper = factory({
      isDone: true,
    });
    const todo = wrapper.find(itemSelector);
    const classes = todo.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });

  it('Date is visible if "dueDate" prop is provided', () => {
    const wrapper = factory({
      dueDate: new Date('03-06-2020'),
    });
    const date = wrapper.find(dateSelector);
    expect(date.exists()).toBe(true);
  });

  it('Return undefined if "dueDate" prop is not provided', () => {
    const wrapper = factory();
    expect(wrapper.vm.getFormattedDate).toEqual(undefined);
  });

  it('Return formatted if "dueDate" prop is provided', () => {
    const wrapper = factory({
      dueDate: new Date('03-06-2020'),
    });
    expect(wrapper.vm.getFormattedDate).not.toEqual(undefined);
  });

  it('Return undefined if "dueDate" prop is not provided', () => {
    const wrapper = factory();
    expect(wrapper.vm.getDeadline).toEqual(undefined);
  });

  it('Return empty string if getDeadline returns undefined', () => {
    const wrapper = factory();
    expect(wrapper.vm.getDeadlineClass).toEqual('');
  });

  it('Return "overdue" if todo date is expired', () => {
    const wrapper = factory({
      dueDate: sub(new Date(), { days: 20 }),
    });
    expect(wrapper.vm.getDeadline).toEqual('overdue');
  });

  it('Return "due-soon" if todo date will expire soon', () => {
    const wrapper = factory({
      dueDate: add(new Date(), { days: 1 }),
    });
    expect(wrapper.vm.getDeadline).toEqual('due-soon');
  });

  it('Return "on-time" if todo date will not expire soon', () => {
    const wrapper = factory({
      dueDate: add(new Date(), { days: 22 }),
    });
    expect(wrapper.vm.getDeadline).toEqual('on-time');
  });
});
