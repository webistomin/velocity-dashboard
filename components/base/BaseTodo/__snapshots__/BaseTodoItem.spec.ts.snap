// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTodoItem Matches snapshot 1`] = `
"<li data-jest=\\"base-todo__item\\" class=\\"list-item base-todo__item \\">
  <div class=\\"base-todo__actions\\"><input data-jest=\\"base-todo__checkbox\\" type=\\"checkbox\\" id=\\"1\\" aria-label=\\"Toggle task with name: Title\\" class=\\"base-todo__checkbox visually-hidden\\" value=\\"1\\"><label for=\\"1\\" class=\\"base-todo__label label\\"></label></div>
  <nuxt-link-stub to=\\"/reminders/1\\" class=\\"base-todo__content link\\"><strong class=\\"base-todo__name\\">Title</strong></nuxt-link-stub>
</li>"
`;
